main/src/library/Iterator.flix

```diff
@@ -216,12 +262,72 @@ namespace Iterator {
     ///
     /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
     ///
-    /// The function `f` must be pure.
+    /// Whether `f` is applied eagerly or lazily depends on its purity:
+    /// - If `f` is pure then it is applied lazily.
+    /// - If `f` is impure then it is applied eagerly (consuming elements of the iterator until and including
+    /// the first element that does *not* satisfy predicate `f`).
     ///
-    /// Consumes elements of the iterator until the first element that does *not* satisfy predicate `f`.
+    /// The original iterator `iter` should not be reused.
+    ///
+    pub def takeWhile(f: a -> Bool & ef, iter: Iterator[a]): Iterator[a] & Impure =
+        matchEff f {
+            case Pure   => takeWhileL(f as a -> Bool, iter)
+            case Impure => takeWhileE(f as a -> Bool & Impure, iter)
+        }
+
+    ///
+    /// Returns the longest prefix of `iter` that satisfies the predicate `f`.
+    /// `f` is applied *lazily*.
+    ///
+    /// Does *not* consume any elements of the iterator.
+    ///
+    /// The original iterator `iter` should not be reused.
+    ///
+    def takeWhileL(f: a -> Bool, iter: Iterator[a]): Iterator[a] & Impure =
+        let Iterator(done, next) = iter;
+        if (done())
+            Iterator(() -> true as & Impure, () -> bug!("Empty iterator") as & Impure)
+        else
+            let x = ref default;
+            let p = () -> {
```

Comment:
Again, a tiny bit more documentation (and/or better name for `p`) would make this quicker to understand.

Reply:
Also, why `p`?

Reply:
`p` was meant as predicate, but I get that it's very vague.